target/snitch_cluster/sw/apps/snax-mac/src/snax-mac.c

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "snrt.h"
+
+#include "data.h"
+
+int main() {
+    // Set err value for checking
+    int err = 0;
+
+    uint32_t final_output;
+
+    uint32_t *local_a, *local_b, *local_c, *local_o;
+
+    // Allocate space in TCDM
+    local_a = (uint32_t *)snrt_l1_next();
+    local_b = local_a + VEC_LEN;
+    local_c = local_b + VEC_LEN;
+    local_o = local_c + 1;
+
+    uint32_t dma_pre_load = snrt_mcycle();
```

Comment:
Have you checked if these variables will be optimized out by `-O3`?

Reply:
Not yet hehe will do later 🛩️

Reply:
@JosseVanDelm where to set this btw?

Reply:
Ah okay, so this was here before already yes? When I observed last time nothing changed much. Wait let me run a trace and see what happens.

Reply:
Nice, I double checked and it does not seem to optimize out or melt. I can still see it but in two different clusters!

For example, in the sample trace where `local_a` up to `local_o` were placed:

```text
0      584        M 0x800001f0 slli    a2, a2, 2              #; a2  = 20, (wrb) a2  <-- 80
0      586        M                                           #; (lsu) a0  <-- 0x10000000
0      587        M 0x800001f4 add     a5, a0, a2             #; a0  = 0x10000000, a2  = 80, (wrb) a5  <-- 0x10000050
0      588        M 0x800001f8 csrr    a3, mcycle             #; mcycle = 587, (wrb) a3  <-- 587
0      589        M 0x800001fc csrr    a3, mhartid            #; mhartid = 0, (wrb) a3  <-- 0
0      590        M 0x80000200 andi    a3, a3, 1              #; a3  = 0, (wrb) a3  <-- 0
0      591        M 0x80000204 add     a6, a5, a2             #; a5  = 0x10000050, a2  = 80, (wrb) a6  <-- 0x100000a0
0      592        M 0x80000208 bnez    a3, pc + 120           #; a3  = 0, not taken
0      593        M 0x8000020c addi    a1, a6, 4              #; a6  = 0x100000a0, (wrb) a1  <-- 0x100000a4
```

Note that TCDM starts at `0x10000000` which is loaded to `local_a`, then 80 bytes later `0x10000050` is `local_b`, then another 80 bytes later that's `0x100000a0` for `local_c` and finally `0x100000a4` for `local_o`. 

So no `-03` did not optimize it 🧨🧨🧨🧨🧨

Reply:
@rgantonio I guess that is because the `read_csr` macro you are using is converted into a `volatile` inline `asm`
https://github.com/pulp-platform/riscv-opcodes/blob/77c3baf1acd8290d7efc8dc94abb659f3b6381c2/encoding.h#L248